internal/controller/controllers/csr_utils.go

```diff
@@ -0,0 +1,228 @@
+			nodeUserPrefix)
+	}
+
+	// Check groups, we need at least:
+	// - system:authenticated
+	if len(csr.Spec.Groups) < 2 || !funk.ContainsString(csr.Spec.Groups, "system:authenticated") {
+		return false, errors.Errorf("CSR %s agent %s/%s: %v is too small or not contains %q", csr.Name, agent.Namespace, agent.Name,
+			csr.Spec.Groups, "system:authenticated")
+	}
+
+	// Check usages, we need only:
+	// - digital signature
+	// - key encipherment
+	// - server auth
+	serverUsages := []certificatesv1.KeyUsage{
+		certificatesv1.UsageDigitalSignature,
+		certificatesv1.UsageKeyEncipherment,
+		certificatesv1.UsageServerAuth,
+	}
+
+	if !hasExactUsages(csr, serverUsages) {
+		return false, errors.Errorf("CSR %s agent %s/%s: No exact match between CSR %v and required %v usages", csr.Name, agent.Namespace, agent.Name,
+			csr.Spec.Usages, serverUsages)
+	}
+	if !funk.ContainsString(x509CSR.Subject.Organization, nodeGroup) {
+		return false, errors.Errorf("CSR %s agent %s/%s: Organization %v doesn't include %s", csr.Name, agent.Namespace, agent.Name,
+			x509CSR.Subject.Organization, nodeGroup)
+	}
+	if x509CSR.Subject.CommonName != csr.Spec.Username {
+		return false, errors.Errorf("CSR %s agent %s/%s: Mismatched CommonName %s != %s for CSR %s", csr.Name, agent.Namespace, agent.Name,
+			x509CSR.Subject.CommonName, csr.Spec.Username, csr.Name)
+	}
+	agentDNSNames := append(getNodeDNSNames(node), getAgentHostname(agent))
+	for _, dnsName := range x509CSR.DNSNames {
+		if !funk.ContainsString(agentDNSNames, dnsName) {
+			return false, errors.Errorf("CSR %s agent %s/%s: DNS name %s missing from available DNS names %v", csr.Name, agent.Namespace, agent.Name,
+				dnsName, agentDNSNames)
+		}
+	}
+	agentIPs := getAgentIPs(agent)
+	for _, ip := range x509CSR.IPAddresses {
+		if !funk.ContainsString(agentIPs, ip.String()) {
+			return false, errors.Errorf("CSR %s agent %s/%s: IP address %s missing from available node IPs %v", csr.Name, agent.Namespace, agent.Name,
+				ip.String(), agentIPs)
+		}
+	}
+	return true, nil
+}
+
+func createNodeServerCsrValidator(node *corev1.Node) nodeCsrValidator {
+	return func(agent *aiv1beta1.Agent, csr *certificatesv1.CertificateSigningRequest, x509cr *x509.CertificateRequest) (bool, error) {
+		return validateNodeServerCSR(agent, node, csr, x509cr)
+	}
+}
+
+func isCsrAssociatedWithAgent(x509CSR *x509.CertificateRequest, agent *aiv1beta1.Agent) bool {
+	return strings.EqualFold(nodeUserPrefix+getAgentHostname(agent), x509CSR.Subject.CommonName)
+}
+
+func isCsrApproved(csr *certificatesv1.CertificateSigningRequest) bool {
+	for _, c := range csr.Status.Conditions {
+		if c.Type == certificatesv1.CertificateApproved {
+			return true
+		}
+	}
+	return false
+}
+
+func inTimeSpan(start, end, check time.Time) bool {
+	return check.After(start) && check.Before(end)
+}
+
+func isCsrInTimeSpan(agent *aiv1beta1.Agent, csr *certificatesv1.CertificateSigningRequest) bool {
```

Comment:
Can you write a small docstring of what this function does and used for, pls

Reply:
OK